torch/onnx/_type_utils.py

```diff
@@ -108,36 +114,94 @@ def from_name(
     @classmethod
     @_beartype.beartype
     def from_dtype(cls, dtype: torch.dtype) -> JitScalarType:
-        """Convert a torch dtype to ScalarType."""
+        """Convert a torch dtype to JitScalarType.
+
+        Note: DO NOT USE this API when `dtype` comes from a `torch._C.Value.type()` calls.
+            A "RuntimeError: INTERNAL ASSERT FAILED at "../aten/src/ATen/core/jit_type_base.h" can
+            be raised in several scenarios where shape info is not present.
+            Instead use `from_value` API which is safer.
+
+        Args:
+            dtype: A torch.dtype to create a JitScalarType from
+
+        Returns:
+            JitScalarType
+
+        Raises:
+            ValueError: if dtype is not a valid torch.dtype or if it is None.
+        """
         if dtype not in _DTYPE_TO_SCALAR_TYPE:
             raise ValueError(f"Unknown dtype: {dtype}")
         return _DTYPE_TO_SCALAR_TYPE[dtype]
 
+    # TODO: `value: torch._C.Value` type annotation raises (CI machines only):
+    #       NameError: name '_C' is not defined
+    # In both cases, using typing.TypeVar or string literals didn't work
+    @classmethod
+    @_beartype.beartype
+    def from_value(cls, value, default=None) -> JitScalarType:
+        """Create a JitScalarType from a torch._C.Value's underlying scalar type.
+
+        Args:
+            value: A `Union[torch._C.Value, torch.Tensor]` object to fetch scalar type from.
+            default: The JitScalarType to return if a valid scalar cannot be fetched from value
+
+        Returns:
+            JitScalarType.
+
+        Raises:
+            OnnxExporterError: if value does not have a valid scalar type and default is not None.
+            SymbolicValueError: when value.type()'s info are empty and default is None
+        """
+
+        if default is None and not isinstance(value, (torch._C.Value, torch.Tensor)):
+            raise errors.OnnxExporterError(
+                "value must be either torch._C.Value, torch.Tensor objects or None."
+            )
+        elif default is not None and not isinstance(default, JitScalarType):
+            raise errors.OnnxExporterError(
+                "default value must be a JitScalarType object or None."
+            )
+        elif isinstance(value, torch.Tensor):
+            return cls.from_dtype(value.dtype)
+        elif isinstance(value.type(), torch.ListType):
+            return cls.from_dtype(value.type().getElementType().dtype())
+        scalar_type = value.type().scalarType()
+        if scalar_type is None:
+            if default is None:
+                raise errors.SymbolicValueError(
+                    f"Cannot determine scalar type for this '{type(value.type())}' instance.",
+                    value,
+                )
+            else:
+                return default
+        return cls.from_name(scalar_type)
```

Comment:
Minor suggestions & quesitions:

1. Removed redundant `elif`s since they are just guards and all branches return https://pylint.pycqa.org/en/latest/user_guide/messages/refactor/no-else-return.html
2. Should we accept None as input to `value`? When do we need this behavior? Would it be better if we tighten the requirement and not accept None?
3. I very much think the first two errors should be TypeError, but I didn't change it in the suggestions below. I feel uneasy relying on manipulating the error type for the fallback behavior, but we can find a better solution in a separate PR.
4. Cleaned up some conditional logic to make them more concise; updated error message for accuracy
5. Updated the control flow in the last branches to reduce nested conditions; raise if every attempt fails

```suggestion
        # TODO: Refine the errors raised to convey a clear intent
        if not isinstance(value, (torch._C.Value, torch.Tensor)):
            raise errors.OnnxExporterError(
                "'value' must be either torch._C.Value or a torch.Tensor object."
            )
        if not isinstance(default, Optional[JitScalarType]):
            raise errors.OnnxExporterError(
                "'default' must be a JitScalarType object or None."
            )
        if isinstance(value, torch.Tensor):
            return cls.from_dtype(value.dtype)
        if isinstance(value.type(), torch.ListType):
            return cls.from_dtype(value.type().getElementType().dtype())

        scalar_type = value.type().scalarType()
        if scalar_type is not None:
            return cls.from_name(scalar_type)
        if default is not None:
            return default

        raise errors.SymbolicValueError(
            f"Cannot determine scalar type for this '{type(value.type())}' instance.",
            value,
        )
```

Reply:
There was a bug in the original checks, `default` is returned even if `value` is bad. fixed